File: go/pkg/iterables/iterables_8.go
package iterables

// ZeroMatrix runs in O(nm)
func ZeroMatrix(m [][]int) {
	zeroRows := map[int]bool{}
	zeroColumns := map[int]bool{}

	for i := 0; i < len(m); i++ {
		for j := 0; j < len(m[0]); j++ {
			if m[i][j] == 0 {
				zeroRows[i] = true
				zeroColumns[j] = true
			}
		}
	}

	for i := 0; i < len(m); i++ {
		for j := 0; j < len(m[0]); j++ {
			if zeroRows[i] || zeroRows[j] {
				m[i][j] = 0
			}
		}
	}
}

File: python/iterables/iterables_8.py
# O(mn) where m and n are the dimensions of the matrix
def zero_matrix(m):
    zero_rows = set()
    zero_columns = set()

    # O(mn)
    for i in range(len(m)):
        for j in range(len(m[0])):
            if m[i][j] == 0:
                zero_rows.add(i)
                zero_columns.add(j)
    # O(mn)
    for i in range(len(m)):
        for j in range(len(m[0])):
            # O(1) lookup by using sets
            if i in zero_rows or j in zero_columns:
                m[i][j] = 0
